File: Silver/2018/February/02_Snow_Boots/SnowBoots2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

/**
 * @author Red
 * @version 1.0
 */

public class SnowBoots2 {
    static List<List<Integer>> boots = new ArrayList<>();
    static int[] snowDepth;
    static int[][] visited;
    static int n, b;

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("snowboots.in"));
        PrintWriter pw = new PrintWriter(new FileWriter("snowboots.out"));
        StringTokenizer st = new StringTokenizer(r.readLine());
        n = Integer.parseInt(st.nextToken());
        b = Integer.parseInt(st.nextToken());
        snowDepth = new int[n];
        visited = new int[n][b];
        st = new StringTokenizer(r.readLine());
        for (int i = 0; i < n; i++) {
            snowDepth[i] = Integer.parseInt(st.nextToken());
        }
        for (int i = 0; i < b; i++) {
            st = new StringTokenizer(r.readLine());
            List<Integer> boot = new ArrayList<>();
            boot.add(Integer.parseInt(st.nextToken()));
            boot.add(Integer.parseInt(st.nextToken()));
            boots.add(boot);
        }
        visited[0][0] = 1;
        // 遍历每个鞋和每个砖块
        for (int boot = 0; boot < b; boot++) {
            int topDepth = boots.get(boot).get(0);
            int topSize = boots.get(boot).get(1);
            for (int location = 0; location < n; location++) {
                // 当前的鞋子无法站住当前的砖块上
                if (snowDepth[location] > topDepth) {
                    visited[location][boot] = -1;
                    continue;
                }
                // 穿着当前的鞋子，前一块可能的砖块
                for (int priorLocation = 0; priorLocation < location; priorLocation++) {
                    if (visited[priorLocation][boot] == 1 && priorLocation + topSize >= location) {
                        visited[location][boot] = 1;
                    }
                }
                // 当前的砖块，前一双可能的鞋子
                for (int priorBoot = 0; priorBoot < boot; priorBoot++) {
                    if (visited[location][priorBoot] == 1) {
                        visited[location][boot] = 1;
                    }
                }
            }
            if (visited[n - 1][boot] == 1) {
                pw.println(boot);
                break;
            }
        }
        pw.close();
    }
}

File: Silver/2018/February/02_snow_boots/snow_boots2.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

vector<vector<int>> boots;
vector<int> snow_depth;
vector<vector<int>> visited;
int n, b;

int main() {
    ifstream fin("snowboots.in");
    ofstream fout("snowboots.out");

    fin >> n >> b;
    snow_depth.resize(n);
    visited.resize(n, vector<int>(b, 0));

    for (int i = 0; i < n; i++) {
        fin >> snow_depth[i];
    }

    for (int i = 0; i < b; i++) {
        vector<int> boot(2);
        fin >> boot[0] >> boot[1];
        boots.push_back(boot);
    }

    visited[0][0] = 1;

    // 遍历每个鞋和每个砖块
    for (int boot = 0; boot < b; boot++) {
        int top_depth = boots[boot][0];
        int top_size = boots[boot][1];

        for (int location = 0; location < n; location++) {
            // 当前的鞋子无法站住当前的砖块上
            if (snow_depth[location] > top_depth) {
                visited[location][boot] = -1;
                continue;
            }

            // 穿着当前的鞋子，前一块可能的砖块
            for (int prior_location = 0; prior_location < location; prior_location++) {
                if (visited[prior_location][boot] == 1 && prior_location + top_size >= location) {
                    visited[location][boot] = 1;
                }
            }

            // 当前的砖块，前一双可能的鞋子
            for (int prior_boot = 0; prior_boot < boot; prior_boot++) {
                if (visited[location][prior_boot] == 1) {
                    visited[location][boot] = 1;
                }
            }
        }

        if (visited[n - 1][boot] == 1) {
            fout << boot << endl;
            break;
        }
    }

    fin.close();
    fout.close();

    return 0;
}
